codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/IdempotencyTokenMiddlewareGenerator.java

```diff
@@ -74,23 +105,47 @@ public void writeAdditionalFiles(
     ) {
-        ServiceShape serviceShape =  settings.getService(model);
+        ServiceShape serviceShape = settings.getService(model);
         Map<ShapeId, MemberShape> map = getOperationsWithIdempotencyToken(model, serviceShape);
         if (map.size() == 0) {
             return;
         }
 
-        writerFactory.accept("idempotencyTokenMiddleware.go", settings.getModuleName(), writer -> {
-            writer.write("var randReader io.Reader = cryptorand.Reader");
-            writer.addUseImports(SmithyGoDependency.IO);
-            for (Map.Entry<ShapeId, MemberShape> entry : map.entrySet()) {
-                ShapeId operationId = entry.getKey();
-                OperationShape operation = model.expectShape(operationId).asOperationShape().get();
-                Shape inputShape  = model.expectShape(operation.getInput().get());
-                MemberShape memberShape = entry.getValue();
-                execute(writer, symbolProvider, operation, memberShape, symbolProvider.toSymbol(inputShape));
-            }
+        writerFactory.accept("api_client.go", settings.getModuleName(), writer -> {
+            writer.write("// IdempotencyTokenProvider interface for providing idempotency token");
+            writer.openBlock("type IdempotencyTokenProvider interface {",
+                    "}", () -> {
+                        writer.write("GetToken() (string, error)");
+                    });
+            writer.write("");
         });
+
+        for (Map.Entry<ShapeId, MemberShape> entry : map.entrySet()) {
+            ShapeId operationShapeId = entry.getKey();
+            OperationShape operation = model.expectShape(operationShapeId, OperationShape.class);
+            writerFactory.accept(String.format("api_op_%s.go", operationShapeId.getName()), settings.getModuleName(),
+                    writer -> {
+                        // Generate idempotency token middleware
+                        MemberShape memberShape = map.get(operationShapeId);
+                        execute(model, writer, symbolProvider, operation, memberShape);
+
+                        // Generate idempotency token middleware registrar function
+                        writer.addUseImports(SmithyGoDependency.SMITHY_MIDDLEWARE);
+                        String middlewareHelperName = getIdempotencyTokenMiddlewareHelperName(operation);
+                        writer.write("// addIdempotencyTokenMiddleware registers middleware into operation stack");
+                        writer.openBlock("func $L("
+                                        + "stack *middleware.Stack, cfg IdempotencyTokenProvider) {",
+                                "}", middlewareHelperName, () -> {
+                                    writer.write("stack.Initialize.Add(&$L{cfg},middleware.After)",
+                                            getIdempotencyTokenMiddlewareName(operation));
+                                });
```

Comment:
tiny nit, the func name won't match the docstring, especially since the generated func is currently being exported.

Reply:
Switched to unexported generated func. Removed doc string as not really required in this case.